File: packages/sdk/src/images-envelope.ts
// The `images-v1` canonical image OUTPUT envelope.
//
// cocore's inference path historically committed `outputCommitment` over
// the UTF-8 bytes of the answer text. Image-generation models produce
// image bytes instead, so a job that emits images commits to a
// CANONICAL-JSON envelope of those images, and the provider's receipt
// sets `outputFormat` to "images-v1" to say so.
//
// The load-bearing property mirrors the `messages-v1` INPUT envelope
// (multimodal-envelope.ts): `outputCommitment` is the SHA-256 over the
// EXACT canonical envelope bytes. The provider builds these bytes and
// hashes them; a verifier holding the logical images reconstructs the
// identical bytes and recomputes the commitment. Neither side parses the
// payload to compute the hash, so it stays self-consistent.
//
// Canonicalization MUST match provider/src/images_envelope.rs (sorted
// keys, no insignificant whitespace) so the Rust provider and this TS
// verifier produce byte-identical output.

import { canonicalBytes } from "./canonical.ts";

/** Wire value for `dev.cocore.compute.receipt.outputFormat` (and the
 *  `dev.cocore.compute.job.outputFormat` hint) when the committed output
 *  bytes are this envelope rather than UTF-8 answer text. */
export const IMAGES_V1 = "images-v1" as const;

/** Schema version carried in the envelope so a verifier can reject a
 *  shape it doesn't understand rather than mis-reading it. */
export const IMAGES_ENVELOPE_VERSION = 1 as const;

/** A single generated image. `data` is the raw image bytes,
 *  base64-encoded, carried inline (no external fetch needed to verify the
 *  commitment). `mime` is the media type (e.g. "image/png"). */
export interface EnvelopeImage {
  mime: string;
  data: string;
}

export interface ImagesEnvelope {
  v: typeof IMAGES_ENVELOPE_VERSION;
  images: EnvelopeImage[];
}

/** Canonical bytes of the images envelope — the exact payload that
 *  `outputCommitment` is computed over. Reuses the shared `canonicalBytes`
 *  so the serialization matches the Rust provider byte for byte. */
export function buildImagesEnvelopeBytes(images: EnvelopeImage[]): Uint8Array {
  const envelope: ImagesEnvelope = { v: IMAGES_ENVELOPE_VERSION, images };
  return canonicalBytes(envelope as unknown as Record<string, unknown>);
}

/** Parse + minimally validate envelope bytes back into the structured
 *  form. Used by verifiers and tests; the provider has its own Rust
 *  parser. Throws on a malformed or unknown-version envelope. */
export function parseImagesEnvelope(bytes: Uint8Array): ImagesEnvelope {
  const obj = JSON.parse(new TextDecoder().decode(bytes)) as unknown;
  if (!obj || typeof obj !== "object") throw new Error("images envelope is not an object");
  const e = obj as Record<string, unknown>;
  if (e.v !== IMAGES_ENVELOPE_VERSION) {
    throw new Error(`unsupported images envelope version: ${String(e.v)}`);
  }
  if (!Array.isArray(e.images)) throw new Error("images envelope.images must be an array");
  const images: EnvelopeImage[] = e.images.map((im, i) => {
    if (!im || typeof im !== "object") throw new Error(`image ${i} is not an object`);
    const img = im as Record<string, unknown>;
    if (typeof img.mime !== "string") throw new Error(`image ${i} mime must be a string`);
    if (typeof img.data !== "string") throw new Error(`image ${i} data must be a string`);
    return { mime: img.mime, data: img.data };
  });
  return { v: IMAGES_ENVELOPE_VERSION, images };
}

File: provider/src/images_envelope.rs
//! Builds the canonical `images-v1` OUTPUT envelope bytes.
//!
//! Mirror of the TS SDK `packages/sdk/src/images-envelope.ts`. An
//! image-generation job commits `outputCommitment` over the SHA-256 of
//! these exact bytes, exactly as the text path commits over the answer's
//! UTF-8 bytes. A requester/verifier holding the delivered images
//! reconstructs the identical bytes here (or in TS) and recomputes the
//! commitment — so the canonicalization MUST stay byte-identical to the
//! SDK (`crate::canonical` already matches `packages/sdk/src/canonical.ts`).
//!
//! Envelope shape: `{ "images": [ { "data": <b64>, "mime": <str> } ], "v": 1 }`
//! (keys sorted, no insignificant whitespace).

use crate::canonical::to_canonical_bytes;
use serde_json::json;

/// One generated image: media type + base64-encoded raw bytes.
#[derive(Debug, Clone)]
pub struct EnvelopeImage {
    pub mime: String,
    pub data_b64: String,
}

/// Canonical bytes of the `images-v1` envelope over `images`, in order.
/// This is the exact payload `outputCommitment` is hashed over.
pub fn build_images_envelope_bytes(images: &[EnvelopeImage]) -> Vec<u8> {
    let arr: Vec<serde_json::Value> = images
        .iter()
        .map(|im| json!({ "mime": im.mime, "data": im.data_b64 }))
        .collect();
    let envelope = json!({ "v": 1, "images": arr });
    // canonical.rs sorts keys + forbids floats; an integer `v` and string
    // fields are always representable, so this cannot fail in practice.
    to_canonical_bytes(&envelope).expect("images envelope canonicalizes")
}

#[cfg(test)]
mod tests {
    use super::*;
    use sha2::{Digest, Sha256};

    // Cross-language parity fixture. The SAME canonical string + SHA-256 are
    // asserted on the TS side (packages/sdk/src/images-envelope.test.ts). A
    // divergence in either canonicalizer is caught here.
    const FIXTURE_CANONICAL: &str = r#"{"images":[{"data":"aGVsbG8=","mime":"image/png"}],"v":1}"#;
    const FIXTURE_SHA256: &str = "87727768fff8767ca20fdc4880a22a2899216722d446d498da395ef1ada58681";

    #[test]
    fn canonical_bytes_match_cross_language_fixture() {
        let images = vec![EnvelopeImage {
            mime: "image/png".into(),
            data_b64: "aGVsbG8=".into(),
        }];
        let bytes = build_images_envelope_bytes(&images);
        assert_eq!(String::from_utf8(bytes.clone()).unwrap(), FIXTURE_CANONICAL);
        let sha = hex::encode(Sha256::digest(&bytes));
        assert_eq!(sha, FIXTURE_SHA256);
    }

    #[test]
    fn preserves_image_order() {
        let images = vec![
            EnvelopeImage {
                mime: "image/png".into(),
                data_b64: "AAAA".into(),
            },
            EnvelopeImage {
                mime: "image/jpeg".into(),
                data_b64: "BBBB".into(),
            },
        ];
        let bytes = build_images_envelope_bytes(&images);
        assert_eq!(
            String::from_utf8(bytes).unwrap(),
            r#"{"images":[{"data":"AAAA","mime":"image/png"},{"data":"BBBB","mime":"image/jpeg"}],"v":1}"#,
        );
    }
}
